File: deno/packages/tools/tools/calendar/google.ts
/**
 * Google Calendar API v3 client.
 *
 * Equivalent to Rust's `brainwires_tools::calendar::google` module.
 */

import type {
  Attendee,
  AttendeeStatus,
  CalendarEvent,
  CalendarInfo,
  FreeBusySlot,
} from "./types.ts";

const CALENDAR_API_BASE = "https://www.googleapis.com/calendar/v3";
const TOKEN_URL = "https://oauth2.googleapis.com/token";

function urlEncode(s: string): string {
  let out = "";
  for (const c of s) {
    if (/[A-Za-z0-9\-_.~]/.test(c)) {
      out += c;
    } else {
      for (const b of new TextEncoder().encode(c)) {
        out += "%" + b.toString(16).padStart(2, "0").toUpperCase();
      }
    }
  }
  return out;
}

interface GoogleTokenResponse {
  access_token: string;
  expires_in?: number;
}

/** Google Calendar API client using OAuth2 credentials. */
export class GoogleCalendarClient {
  private access_token: string;
  private readonly client_id: string;
  private readonly client_secret: string;
  private readonly refresh_token: string;

  private constructor(
    access_token: string,
    client_id: string,
    client_secret: string,
    refresh_token: string,
  ) {
    this.access_token = access_token;
    this.client_id = client_id;
    this.client_secret = client_secret;
    this.refresh_token = refresh_token;
  }

  /** Create a new client and obtain an access token via refresh. */
  static async create(
    client_id: string,
    client_secret: string,
    refresh_token: string,
  ): Promise<GoogleCalendarClient> {
    const token = await GoogleCalendarClient.refreshAccessToken(
      client_id,
      client_secret,
      refresh_token,
    );
    return new GoogleCalendarClient(
      token,
      client_id,
      client_secret,
      refresh_token,
    );
  }

  private static async refreshAccessToken(
    client_id: string,
    client_secret: string,
    refresh_token: string,
  ): Promise<string> {
    const body = new URLSearchParams({
      client_id,
      client_secret,
      refresh_token,
      grant_type: "refresh_token",
    }).toString();
    const resp = await fetch(TOKEN_URL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body,
    });
    if (!resp.ok) {
      throw new Error(
        `Failed to refresh Google OAuth2 token: ${resp.status}`,
      );
    }
    const json = await resp.json() as GoogleTokenResponse;
    return json.access_token;
  }

  /** Re-authenticate and update the access token. */
  async ensureToken(): Promise<void> {
    this.access_token = await GoogleCalendarClient.refreshAccessToken(
      this.client_id,
      this.client_secret,
      this.refresh_token,
    );
  }

  /** List events from a calendar within a time range. */
  async listEvents(
    calendar_id: string,
    time_min: string | null | undefined,
    time_max: string | null | undefined,
    max_results: number,
  ): Promise<CalendarEvent[]> {
    let url = `${CALENDAR_API_BASE}/calendars/${urlEncode(calendar_id)}/events` +
      `?maxResults=${max_results}&singleEvents=true&orderBy=startTime`;
    if (time_min) url += `&timeMin=${urlEncode(time_min)}`;
    if (time_max) url += `&timeMax=${urlEncode(time_max)}`;

    const resp = await fetch(url, {
      headers: { Authorization: `Bearer ${this.access_token}` },
    });
    if (!resp.ok) {
      throw new Error(`Failed to list Google Calendar events: ${resp.status}`);
    }
    const body = await resp.json() as { items?: unknown[] };
    const items = Array.isArray(body.items) ? body.items : [];
    const events: CalendarEvent[] = [];
    for (const item of items) {
      const e = GoogleCalendarClient.parseEvent(item);
      if (e) events.push(e);
    }
    return events;
  }

  /** Create a new calendar event. */
  async createEvent(
    calendar_id: string,
    event: CalendarEvent,
  ): Promise<CalendarEvent> {
    const url =
      `${CALENDAR_API_BASE}/calendars/${urlEncode(calendar_id)}/events`;
    const body = GoogleCalendarClient.eventToGoogleJson(event);
    const resp = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.access_token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    if (!resp.ok) {
      throw new Error(
        `Failed to create Google Calendar event: ${resp.status}`,
      );
    }
    const resbody = await resp.json();
    const created = GoogleCalendarClient.parseEvent(resbody);
    if (!created) throw new Error("Failed to parse created event response");
    return created;
  }

  /** Update an existing calendar event. */
  async updateEvent(
    calendar_id: string,
    event_id: string,
    event: CalendarEvent,
  ): Promise<CalendarEvent> {
    const url = `${CALENDAR_API_BASE}/calendars/${urlEncode(calendar_id)}` +
      `/events/${urlEncode(event_id)}`;
    const body = GoogleCalendarClient.eventToGoogleJson(event);
    const resp = await fetch(url, {
      method: "PUT",
      headers: {
        Authorization: `Bearer ${this.access_token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    if (!resp.ok) {
      throw new Error(
        `Failed to update Google Calendar event: ${resp.status}`,
      );
    }
    const resbody = await resp.json();
    const updated = GoogleCalendarClient.parseEvent(resbody);
    if (!updated) throw new Error("Failed to parse updated event response");
    return updated;
  }

  /** Delete a calendar event. */
  async deleteEvent(calendar_id: string, event_id: string): Promise<void> {
    const url = `${CALENDAR_API_BASE}/calendars/${urlEncode(calendar_id)}` +
      `/events/${urlEncode(event_id)}`;
    const resp = await fetch(url, {
      method: "DELETE",
      headers: { Authorization: `Bearer ${this.access_token}` },
    });
    if (!resp.ok && resp.status !== 204) {
      const body = await resp.text().catch(() => "");
      throw new Error(`Delete failed: ${body}`);
    }
  }

  /** Query free/busy information. */
  async freeBusy(
    calendar_ids: readonly string[],
    time_min: string,
    time_max: string,
  ): Promise<FreeBusySlot[]> {
    const url = `${CALENDAR_API_BASE}/freeBusy`;
    const items = calendar_ids.map((id) => ({ id }));
    const body = {
      timeMin: time_min,
      timeMax: time_max,
      items,
    };
    const resp = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.access_token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    if (!resp.ok) {
      throw new Error(`Failed to query free/busy: ${resp.status}`);
    }
    const res = await resp.json() as {
      calendars?: Record<string, { busy?: Array<{ start?: string; end?: string }> }>;
    };
    const slots: FreeBusySlot[] = [];
    if (res.calendars) {
      for (const cal of Object.values(res.calendars)) {
        for (const busy of cal.busy ?? []) {
          slots.push({
            start: busy.start ?? "",
            end: busy.end ?? "",
            status: "busy",
          });
        }
      }
    }
    return slots;
  }

  /** List available calendars. */
  async listCalendars(): Promise<CalendarInfo[]> {
    const url = `${CALENDAR_API_BASE}/users/me/calendarList`;
    const resp = await fetch(url, {
      headers: { Authorization: `Bearer ${this.access_token}` },
    });
    if (!resp.ok) throw new Error(`Failed to list calendars: ${resp.status}`);
    const body = await resp.json() as { items?: unknown[] };
    const items = Array.isArray(body.items) ? body.items : [];
    return items.map((item) => {
      const raw = item as Record<string, unknown>;
      return {
        id: (raw.id as string | undefined) ?? "",
        name: (raw.summary as string | undefined) ?? "",
        color: (raw.backgroundColor as string | undefined) ?? null,
        primary: raw.primary === true,
      };
    });
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  /** Parse a Google Calendar event object into a CalendarEvent. */
  static parseEvent(item: unknown): CalendarEvent | null {
    if (!item || typeof item !== "object") return null;
    const raw = item as Record<string, unknown>;
    const id = typeof raw.id === "string" ? raw.id : null;
    if (!id) return null;
    const title = typeof raw.summary === "string" ? raw.summary : "";

    const startObj = raw.start as Record<string, unknown> | undefined;
    let start = "";
    let all_day = false;
    if (startObj) {
      if (typeof startObj.dateTime === "string") {
        start = startObj.dateTime;
      } else if (typeof startObj.date === "string") {
        start = startObj.date;
        all_day = true;
      } else {
        return null;
      }
    } else {
      return null;
    }

    const endObj = raw.end as Record<string, unknown> | undefined;
    let end = start;
    if (endObj) {
      if (typeof endObj.dateTime === "string") {
        end = endObj.dateTime;
      } else if (typeof endObj.date === "string") {
        end = endObj.date;
      }
    }

    const attendeesRaw = Array.isArray(raw.attendees) ? raw.attendees : [];
    const attendees: Attendee[] = [];
    for (const a of attendeesRaw) {
      if (!a || typeof a !== "object") continue;
      const ar = a as Record<string, unknown>;
      const email = typeof ar.email === "string" ? ar.email : null;
      if (!email) continue;
      const name = typeof ar.displayName === "string" ? ar.displayName : null;
      const rs = typeof ar.responseStatus === "string"
        ? ar.responseStatus
        : "needsAction";
      let status: AttendeeStatus;
      switch (rs) {
        case "accepted":
          status = "accepted";
          break;
        case "declined":
          status = "declined";
          break;
        case "tentative":
          status = "tentative";
          break;
        default:
          status = "needsAction";
      }
      attendees.push({ email, name, status });
    }

    return {
      id,
      title,
      description: typeof raw.description === "string" ? raw.description : null,
      location: typeof raw.location === "string" ? raw.location : null,
      start,
      end,
      all_day,
      attendees,
      recurrence: null,
      reminders: [],
      calendar_id: null,
    };
  }

  /** Convert a CalendarEvent into the Google Calendar JSON shape. */
  static eventToGoogleJson(event: CalendarEvent): Record<string, unknown> {
    const startObj = event.all_day
      ? { date: event.start }
      : { dateTime: event.start };
    const endObj = event.all_day
      ? { date: event.end }
      : { dateTime: event.end };

    const body: Record<string, unknown> = {
      summary: event.title,
      start: startObj,
      end: endObj,
    };
    if (event.description) body.description = event.description;
    if (event.location) body.location = event.location;
    if (event.attendees.length > 0) {
      body.attendees = event.attendees.map((a) => {
        const obj: Record<string, unknown> = { email: a.email };
        if (a.name) obj.displayName = a.name;
        return obj;
      });
    }
    return body;
  }
}

File: crates/brainwires-tool-builtins/src/calendar/google.rs
//! Google Calendar API v3 client.

use anyhow::{Context, Result};
use reqwest::Client;
use serde::Deserialize;
use serde_json::Value;

use super::types::{
    Attendee, AttendeeStatus, BusyStatus, CalendarEvent, CalendarInfo, FreeBusySlot,
};

const CALENDAR_API_BASE: &str = "https://www.googleapis.com/calendar/v3";
const TOKEN_URL: &str = "https://oauth2.googleapis.com/token";

/// Google Calendar API client using OAuth2 credentials.
pub struct GoogleCalendarClient {
    client: Client,
    access_token: String,
    client_id: String,
    client_secret: String,
    refresh_token: String,
}

#[derive(Deserialize)]
struct TokenResponse {
    access_token: String,
    #[allow(dead_code)]
    expires_in: u64,
}

impl GoogleCalendarClient {
    /// Create a new client and obtain an access token via refresh.
    pub async fn new(client_id: &str, client_secret: &str, refresh_token: &str) -> Result<Self> {
        let client = Client::new();
        let access_token =
            Self::refresh_access_token(&client, client_id, client_secret, refresh_token).await?;

        Ok(Self {
            client,
            access_token,
            client_id: client_id.to_string(),
            client_secret: client_secret.to_string(),
            refresh_token: refresh_token.to_string(),
        })
    }

    /// Refresh the OAuth2 access token.
    async fn refresh_access_token(
        client: &Client,
        client_id: &str,
        client_secret: &str,
        refresh_token: &str,
    ) -> Result<String> {
        let resp: TokenResponse = client
            .post(TOKEN_URL)
            .form(&[
                ("client_id", client_id),
                ("client_secret", client_secret),
                ("refresh_token", refresh_token),
                ("grant_type", "refresh_token"),
            ])
            .send()
            .await
            .context("Failed to refresh Google OAuth2 token")?
            .json()
            .await
            .context("Failed to parse token response")?;

        Ok(resp.access_token)
    }

    /// Re-authenticate if the current token has expired.
    pub async fn ensure_token(&mut self) -> Result<()> {
        // Simple strategy: always refresh. A production implementation would
        // check expiry time and only refresh when needed.
        self.access_token = Self::refresh_access_token(
            &self.client,
            &self.client_id,
            &self.client_secret,
            &self.refresh_token,
        )
        .await?;
        Ok(())
    }

    /// List events from a calendar within a time range.
    pub async fn list_events(
        &self,
        calendar_id: &str,
        time_min: Option<&str>,
        time_max: Option<&str>,
        max_results: u32,
    ) -> Result<Vec<CalendarEvent>> {
        let mut url = format!(
            "{}/calendars/{}/events?maxResults={}&singleEvents=true&orderBy=startTime",
            CALENDAR_API_BASE,
            urlencoding::encode(calendar_id),
            max_results,
        );
        if let Some(min) = time_min {
            url.push_str(&format!("&timeMin={}", urlencoding::encode(min)));
        }
        if let Some(max) = time_max {
            url.push_str(&format!("&timeMax={}", urlencoding::encode(max)));
        }

        let resp: Value = self
            .client
            .get(&url)
            .bearer_auth(&self.access_token)
            .send()
            .await
            .context("Failed to list Google Calendar events")?
            .json()
            .await?;

        let items = resp
            .get("items")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let events = items.iter().filter_map(Self::parse_event).collect();
        Ok(events)
    }

    /// Create a new calendar event.
    pub async fn create_event(
        &self,
        calendar_id: &str,
        event: &CalendarEvent,
    ) -> Result<CalendarEvent> {
        let url = format!(
            "{}/calendars/{}/events",
            CALENDAR_API_BASE,
            urlencoding::encode(calendar_id),
        );

        let body = Self::event_to_google_json(event);

        let resp: Value = self
            .client
            .post(&url)
            .bearer_auth(&self.access_token)
            .json(&body)
            .send()
            .await
            .context("Failed to create Google Calendar event")?
            .json()
            .await?;

        Self::parse_event(&resp)
            .ok_or_else(|| anyhow::anyhow!("Failed to parse created event response"))
    }

    /// Update an existing calendar event.
    pub async fn update_event(
        &self,
        calendar_id: &str,
        event_id: &str,
        event: &CalendarEvent,
    ) -> Result<CalendarEvent> {
        let url = format!(
            "{}/calendars/{}/events/{}",
            CALENDAR_API_BASE,
            urlencoding::encode(calendar_id),
            urlencoding::encode(event_id),
        );

        let body = Self::event_to_google_json(event);

        let resp: Value = self
            .client
            .put(&url)
            .bearer_auth(&self.access_token)
            .json(&body)
            .send()
            .await
            .context("Failed to update Google Calendar event")?
            .json()
            .await?;

        Self::parse_event(&resp)
            .ok_or_else(|| anyhow::anyhow!("Failed to parse updated event response"))
    }

    /// Delete a calendar event.
    pub async fn delete_event(&self, calendar_id: &str, event_id: &str) -> Result<()> {
        let url = format!(
            "{}/calendars/{}/events/{}",
            CALENDAR_API_BASE,
            urlencoding::encode(calendar_id),
            urlencoding::encode(event_id),
        );

        let resp = self
            .client
            .delete(&url)
            .bearer_auth(&self.access_token)
            .send()
            .await
            .context("Failed to delete Google Calendar event")?;

        if resp.status().is_success() || resp.status().as_u16() == 204 {
            Ok(())
        } else {
            let body = resp.text().await.unwrap_or_default();
            anyhow::bail!("Delete failed: {}", body)
        }
    }

    /// Query free/busy information.
    pub async fn free_busy(
        &self,
        calendar_ids: &[String],
        time_min: &str,
        time_max: &str,
    ) -> Result<Vec<FreeBusySlot>> {
        let url = format!("{}/freeBusy", CALENDAR_API_BASE);

        let items: Vec<Value> = calendar_ids
            .iter()
            .map(|id| serde_json::json!({"id": id}))
            .collect();

        let body = serde_json::json!({
            "timeMin": time_min,
            "timeMax": time_max,
            "items": items,
        });

        let resp: Value = self
            .client
            .post(&url)
            .bearer_auth(&self.access_token)
            .json(&body)
            .send()
            .await
            .context("Failed to query free/busy")?
            .json()
            .await?;

        let mut slots = Vec::new();
        if let Some(calendars) = resp.get("calendars").and_then(|c| c.as_object()) {
            for (_cal_id, cal_data) in calendars {
                if let Some(busy_arr) = cal_data.get("busy").and_then(|b| b.as_array()) {
                    for slot in busy_arr {
                        let start = slot
                            .get("start")
                            .and_then(|s| s.as_str())
                            .unwrap_or_default();
                        let end = slot.get("end").and_then(|e| e.as_str()).unwrap_or_default();
                        slots.push(FreeBusySlot {
                            start: start.to_string(),
                            end: end.to_string(),
                            status: BusyStatus::Busy,
                        });
                    }
                }
            }
        }
        Ok(slots)
    }

    /// List available calendars.
    pub async fn list_calendars(&self) -> Result<Vec<CalendarInfo>> {
        let url = format!("{}/users/me/calendarList", CALENDAR_API_BASE);

        let resp: Value = self
            .client
            .get(&url)
            .bearer_auth(&self.access_token)
            .send()
            .await
            .context("Failed to list calendars")?
            .json()
            .await?;

        let items = resp
            .get("items")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let calendars = items
            .iter()
            .map(|item| CalendarInfo {
                id: item
                    .get("id")
                    .and_then(|v| v.as_str())
                    .unwrap_or_default()
                    .to_string(),
                name: item
                    .get("summary")
                    .and_then(|v| v.as_str())
                    .unwrap_or_default()
                    .to_string(),
                color: item
                    .get("backgroundColor")
                    .and_then(|v| v.as_str())
                    .map(String::from),
                primary: item
                    .get("primary")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false),
            })
            .collect();

        Ok(calendars)
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    fn parse_event(item: &Value) -> Option<CalendarEvent> {
        let id = item.get("id")?.as_str()?.to_string();
        let title = item
            .get("summary")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_string();

        let (start, all_day) = if let Some(s) = item.get("start") {
            if let Some(dt) = s.get("dateTime").and_then(|v| v.as_str()) {
                (dt.to_string(), false)
            } else if let Some(d) = s.get("date").and_then(|v| v.as_str()) {
                (d.to_string(), true)
            } else {
                return None;
            }
        } else {
            return None;
        };

        let end = if let Some(e) = item.get("end") {
            if let Some(dt) = e.get("dateTime").and_then(|v| v.as_str()) {
                dt.to_string()
            } else if let Some(d) = e.get("date").and_then(|v| v.as_str()) {
                d.to_string()
            } else {
                start.clone()
            }
        } else {
            start.clone()
        };

        let attendees = item
            .get("attendees")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|a| {
                        let email = a.get("email")?.as_str()?.to_string();
                        let name = a
                            .get("displayName")
                            .and_then(|v| v.as_str())
                            .map(String::from);
                        let status = match a
                            .get("responseStatus")
                            .and_then(|v| v.as_str())
                            .unwrap_or("needsAction")
                        {
                            "accepted" => AttendeeStatus::Accepted,
                            "declined" => AttendeeStatus::Declined,
                            "tentative" => AttendeeStatus::Tentative,
                            _ => AttendeeStatus::NeedsAction,
                        };
                        Some(Attendee {
                            email,
                            name,
                            status,
                        })
                    })
                    .collect()
            })
            .unwrap_or_default();

        Some(CalendarEvent {
            id,
            title,
            description: item
                .get("description")
                .and_then(|v| v.as_str())
                .map(String::from),
            location: item
                .get("location")
                .and_then(|v| v.as_str())
                .map(String::from),
            start,
            end,
            all_day,
            attendees,
            recurrence: None,
            reminders: vec![],
            calendar_id: None,
        })
    }

    fn event_to_google_json(event: &CalendarEvent) -> Value {
        let start = if event.all_day {
            serde_json::json!({"date": event.start})
        } else {
            serde_json::json!({"dateTime": event.start})
        };
        let end = if event.all_day {
            serde_json::json!({"date": event.end})
        } else {
            serde_json::json!({"dateTime": event.end})
        };

        let mut body = serde_json::json!({
            "summary": event.title,
            "start": start,
            "end": end,
        });

        if let Some(ref desc) = event.description {
            body["description"] = serde_json::json!(desc);
        }
        if let Some(ref loc) = event.location {
            body["location"] = serde_json::json!(loc);
        }
        if !event.attendees.is_empty() {
            let attendees: Vec<Value> = event
                .attendees
                .iter()
                .map(|a| {
                    let mut obj = serde_json::json!({"email": a.email});
                    if let Some(ref name) = a.name {
                        obj["displayName"] = serde_json::json!(name);
                    }
                    obj
                })
                .collect();
            body["attendees"] = serde_json::json!(attendees);
        }

        body
    }
}

/// Encode a string for use in URL paths (re-exported for convenience).
mod urlencoding {
    pub fn encode(s: &str) -> String {
        url_encode(s)
    }

    fn url_encode(s: &str) -> String {
        let mut result = String::with_capacity(s.len());
        for c in s.chars() {
            match c {
                'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => result.push(c),
                _ => {
                    for b in c.to_string().as_bytes() {
                        result.push_str(&format!("%{:02X}", b));
                    }
                }
            }
        }
        result
    }
}
